packages/buidler-core/src/internal/core/errors.ts

```diff
@@ -9,75 +8,181 @@ export interface ErrorDescription {
   message: string;
 }
 
-export function isBuidlerError(error: any): error is BuidlerError {
-  return error.isBuidlerError;
-}
-
-export function isBuidlerPluginError(error: any): error is BuidlerPluginError {
-  return error.isBuidlerPluginError;
-}
+// For an explanation about these classes constructors go to:
+// https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
 
 export class BuidlerError extends Error {
   public readonly number: number;
   public readonly parent?: Error;
-  public readonly isBuidlerError: true;
 
   constructor(
     errorDescription: ErrorDescription,
     parentError: Error,
-    ...messageArguments: any[]
+    messageArguments?: { [variableName: string]: any }
+  );
+  constructor(
+    errorDescription: ErrorDescription,
+    messageArguments?: { [variableName: string]: any }
   );
-  constructor(errorDescription: ErrorDescription, ...messageArguments: any[]);
   constructor(
     errorDescription: ErrorDescription,
-    parentError?: Error,
-    ...messageArguments: any[]
+    parentError: Error | { [variableName: string]: any } = {},
```

Comment:
I would make two changes to this:

1. Take the params in this order: `(errorDescription, messageArguments, parentError)`, because most of the time you'll pass in some arguments, but not the parent.

2. With the above change, simply the types to the following:
  ```typescript
(errorDescription: ErrorDescription, messageArguments: { [var: string]: string }, parentError: Error | undefined)
  ```
  Notice I also changed the `any` in the arguments map to `string`. This improves type safety, we ensure that users correctly convert their arguments to strings before interpolating them in the message.

Reply:
Thanks for this suggestion, I like it. `BuidlerError` is internal, so we can change it :)

About the `any`:  `applyErrorMessageTemplate` handles non-string values. An initial iteration only accepted `string`s, but that lead to code repetition without much of an upside.